presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -172,10 +173,12 @@ protected InternalResourceGroup(
         if (parent.isPresent()) {
             id = new ResourceGroupId(parent.get().id, name);
             root = parent.get().root;
+            this.lock = parent.get().lock;
         }
         else {
             id = new ResourceGroupId(name);
             root = this;
+            this.lock = new ReentrantReadWriteLock();
```

Comment:
> Adding some unit test to make sure the methods fail when the thread has not acquired a lock beforehand is great. (i.e. around internalStartNext, startInBackground, enqueueQuery etc).
> 
> Also for the the shadow run, what exactly we tested? The test should cover the cluster with ~2K queries on it (running + queued), we should check if adjusted queue size metric is still publishing data and not timing out.

My initial shadow run was mainly to make sure there are no deadlock introduced. I will do more testing with both fair and non-fair mode. I was a bit worried to use fair mode as there is an impact on throughput, but I will do more testing to verify this.

Reply:
Based on testing and also digging deeper into the code, it looks like reader-writer lock is not going to help much here as we have more write locks compared to read locks when the query throughput is high. Also based on testing, metrics (not just adjusted queue size, but all other metrics) are stopped getting published when cpu utilization is consistently ~83% for longer time, so the issue seems to be not due to lock contention. I have converted this PR to draft for future reference in case we want to optimize locking further (places where we feel we can downgrade locking from write to read).